core/arena/unique_arena.rs

```diff
@@ -0,0 +1,359 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+use std::alloc::Layout;
+use std::future::Future;
+use std::pin::Pin;
+use std::ptr::NonNull;
+
+use crate::arena::raw_arena::RawArena;
+
+use super::alloc;
+
+#[cfg(debug_assertions)]
+const SIGNATURE: usize = 0x8877665544332211;
+
+pub struct ArenaUniqueReservation<T>(NonNull<ArenaBoxData<T>>);
+
+impl<T> Drop for ArenaUniqueReservation<T> {
+  fn drop(&mut self) {
+    panic!("A reservation must be completed or forgotten")
+  }
+}
+
+pub struct ArenaBox<T: 'static> {
+  ptr: NonNull<ArenaBoxData<T>>,
+}
+
+impl<T> Unpin for ArenaBox<T> {}
+
+struct ArenaBoxData<T> {
+  #[cfg(debug_assertions)]
+  signature: usize,
+  arena_data: NonNull<ArenaUniqueData<T>>,
+  data: T,
+}
+
+impl<T: 'static> ArenaBox<T> {
+  /// Offset of the `ptr` field within the `ArenaBox` struct.
+  const PTR_OFFSET: usize = memoffset::offset_of!(ArenaBox<T>, ptr);
```

Comment:
I wonder if instead of using this crate, we could use newly stabilized APIs from the standard library? https://blog.rust-lang.org/2023/12/28/Rust-1.75.0.html#pointer-byte-offset-apis

Reply:
I didn't see those new APIs, sweet. We still need the memoffset crate to get the offsets, but we can simplify the pointer math below.

Reply:
I'm actually going to punt on this one until we're certain of how reliable 1.75 is -- the APIs don't buy us a lot right now and I want to avoid locking in 1.75 right now. It'll be easy enough to fix with a future revision.

Reply:
In the end I moved all the byte-manipulation functions to one spot to make it easier to swap out in the future